internal/releasesources/clients.go

```diff
@@ -0,0 +1,146 @@
+type OpenShiftReleasesAPIClientInterface interface {
+	getEndpoint(channel, openshiftVersion, cpuArchitecture string) string
+	parseResponse(response *http.Response) (*releaseGraph, error)
+	GetReleases(channel, openshiftVersion, cpuArchitecture string) (*releaseGraph, error)
+}
+
+//go:generate mockgen -source=clients.go -destination=mock_clients.go -package=releasesources
+type OpenShiftSupportLevelAPIClientInterface interface {
+	getEndpoint(openshiftMajorVersion string) string
+	parseResponse(response *http.Response) (*supportLevelGraph, error)
+	createSupportLevelsMap(supportLevelGraph *supportLevelGraph) OCPVersionSupportLevels
+	GetSupportLevels(majorVersion string) (OCPVersionSupportLevels, error)
+}
+
+type releaseGraph struct {
+	Nodes []node `json:"nodes"`
+}
+
+type node struct {
+	Version string `json:"version"`
+}
+
+type supportLevelGraph struct {
+	Data []data `json:"data"`
+}
+
+type data struct {
+	Versions []version `json:"versions"`
+}
+
+type version struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+}
+
+const releasesAPIQueryTemplate string = "%s?channel=%s-%s&arch=%s"
+const supportLevelAPIQueryTemplate string = "%s?name=Openshift+Container+Platform+%s"
+
+type OpenShiftReleasesAPIClient struct {
+	BaseUrl string
+}
+
+type OpenShiftSupportLevelAPIClient struct {
+	BaseUrl string
+}
+
+func createOCPReleaseAPIBaseEndpoint(apiEndpoint string) string {
+	return fmt.Sprintf("%s/%s", apiEndpoint, releaseImagesResourcePath)
+}
+
+func (o OpenShiftReleasesAPIClient) getEndpoint(channel, openshiftVersion, cpuArchitecture string) string {
+	return fmt.Sprintf(releasesAPIQueryTemplate, o.BaseUrl, channel, openshiftVersion, cpuArchitecture)
+}
+
+func (o OpenShiftReleasesAPIClient) parseResponse(response *http.Response) (*releaseGraph, error) {
+	var graph releaseGraph
+	err := json.NewDecoder(response.Body).Decode(&graph)
+	if err != nil {
+		return nil, err
+	}
+
+	return &graph, nil
+}
+
+func (o OpenShiftReleasesAPIClient) GetReleases(channel, openshiftVersion, cpuArchitecture string) (*releaseGraph, error) {
+	endpoint := o.getEndpoint(channel, openshiftVersion, cpuArchitecture)
+	response, err := http.Get(endpoint)
+	if err != nil {
+		return nil, errors.Errorf("an error occurred while making http request to %s: %s", endpoint, err)
+	}
+
+	graph, err := o.parseResponse(response)
+	if err != nil {
+		return nil, errors.Errorf("an error occurred while decoding the response to a request made to %s: %s", endpoint, err)
+	}
+
+	return graph, nil
+}
+
+func (o OpenShiftSupportLevelAPIClient) getEndpoint(openshiftMajorVersion string) string {
+	return fmt.Sprintf(supportLevelAPIQueryTemplate, o.BaseUrl, openshiftMajorVersion)
+}
+
+func (o OpenShiftSupportLevelAPIClient) parseResponse(response *http.Response) (*supportLevelGraph, error) {
+	var graph supportLevelGraph
+	err := json.NewDecoder(response.Body).Decode(&graph)
+	if err != nil {
+		return nil, err
```

Comment:
Lets add more description to err

Reply:
Isn't the description for the error in `GetReleases` sufficient ?